docs/ROADMAP.md

```diff
@@ -0,0 +1,119 @@
+      [Spike](https://github.com/strongloop/loopback-next/issues/2027)
+- [ ] Stretch goal: enable class level `@authenticate` decorator:
+      https://github.com/strongloop/loopback-next/issues/2460
+
+#### Authorization
+
+- [ ] Complete the `Add authorization component` PR:
+      https://github.com/strongloop/loopback.io/pull/857
+- [ ] Common layer for authentication & authorization
+      https://github.com/strongloop/loopback-next/issues/2900
+- [ ] Add authorization to example-shopping repo
+
+#### Inclusion of Related Models https://github.com/strongloop/loopback-next/issues/1352
+
+- awaiting to @bajtos' spike to create the tasks
+
+#### From model definition to REST API with no custom repository/controller classes (Declarative Support) https://github.com/strongloop/loopback-next/issues/2036
+
+- [ ] [CRUDRESTController](https://github.com/strongloop/loopback-next/issues/2736)
+- [ ] [KeyValueRESTController](https://github.com/strongloop/loopback-next/issues/2737)
+- [ ] [Spike: Booter for creating REST APIs from model files](https://github.com/strongloop/loopback-next/issues/2738)
+
+#### CI cleanup https://github.com/strongloop-internal/scrum-apex/issues/422
+
+- [ ] Fix CI for
+      [dashdb](https://github.com/strongloop/loopback-connector-dashdb/issues/76)
+- [ ] Fix CI for
+      [db2](https://github.com/strongloop/loopback-connector-db2/issues/130)
+      connector
+- [ ] Stretch goal: Test juggler 3.x and 4.x on connectors
+
+#### Support of Node.js 12 https://github.com/strongloop/loopback-next/issues/3072
+
+- [ ] [SQL connectors](https://github.com/strongloop/loopback-next/issues/3110)
+- [ ] [noSQL connectors](https://github.com/strongloop/loopback-next/issues/3111)
+- [ ] [service connectors](https://github.com/strongloop/loopback-next/issues/3112)
+
+#### Internal tooling - to address build time
+
+- Spike: Investigate setting up Windows CI workflow for connectors using Azure
+  Pipelines, https://github.com/strongloop/loopback-next/issues/3161
+
+#### Documentation improvement
+
+- Tasks created according to this proposal:
+  https://github.com/strongloop/loopback-next/pull/2925
+
+#### Miscellaneous
+
+- Review of
+  [DB2 connector on IBM i](https://github.com/strongloop/loopback-connector-ibmi)
+
+#### Bug fixes / Developer Experience Improvements
+
+- Bugs:
+  https://github.com/strongloop/loopback-next/issues?q=is%3Aopen+is%3Aissue+label%3Abug+-label%3A%22good+first+issue%22+-label%3A%22help+wanted%22
+
+- Developer Experience (Pick a few):
+  https://github.com/strongloop/loopback-next/issues?utf8=✓&q=is%3Aopen+is%3Aissue+-label%3A%22good+first+issue%22+-label%3A%22help+wanted%22+label%3Adeveloper-experience+
+
+#### Juggler next
+
+- spike to determine where the code should reside,
+  https://github.com/strongloop/loopback-next/issues/3126
+
+### Stretch Goals
+
+---
+
+#### Juggler next
+
+- Convert base repo (e.g. loopback-connector) and more popular connectors to
+  TypeScript and to use async/await
```

Comment:
I'd like to propose a slightly different approach.

1. Improve current juggler v4 to allow connectors to implement DAO/KVAO methods as async functions returning a promise (not accepting a callback).
2. Rework `loopback-connector` from ES5 codebase to ES2017 (`class` keyword, `async` functions). This is a breaking change for connectors consuming this module, but should be done in backwards-compatible way from the point of view of LB4 applications. As part of this change, we can migrate the code to TypeScript, although I would personally prefer to leave TypeScript for later. It's already more than enough work to upgrade from ES5 to ES2017.
3. Upgrade most (if not all) of our connectors to use the new major version of `loopback-connector`, leverage class inheritance and async functions.

Reply:
updated. I'd leave the migrating to TS part for now as you mentioned it's a big effort to do that.